### PR TITLE
ACAS-633: Google cloud storage and creg bulkloader subfolders

### PR DESCRIPTION
**Pull request by brianbolt**
## Description
 - Implement GCS file handler (not just local storage)
 - Store BulkLoader files in subfolders with id of bulk load
 

## Related Issue
ACAS-633

## How Has This Been Tested?
Manual UI testing to confirm
  - Temporary uploaded files could be uploaded and downloaded
  - Stored files such as Experiment files, Bulkloader files could be downloaded from the UI
Added acasclient tests for report file and image file checking (see ACAS-633 acasclient pr
Ran acasclient tests when using GCS

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1 +1 @@
-ACAS_TAG=latest
\ No newline at end of file
+ACAS_TAG=release-2023.2.x
```
Comment (brianbolt) [inline .env:1]: Plan on reverting before merging

File: conf/config.properties.example [hunk 2]
```diff
@@ -322,9 +322,23 @@ server.delete.files.on.reload=false
 server.projects.type=HardCoded
 client.require.login=true
 
+# blueimp for local file storage
+# gcs for Google Cloud Storage
+# custom for custom file storage (add file service url below)
 server.service.external.file.type=blueimp
+server.service.external.file.uploadTempFiles=true
+server.migrateCmpdRegBulkLoaderFilesToSubfoldersOnStart=true
```
Comment (brianbolt) [inline conf/config.properties.example:330]: Plan on leaving this as a default and then switching to false in a few quarters so its not run on every start of ACAS.

File: docker-compose.yml [hunk 3]
```diff
@@ -35,7 +35,7 @@ services:
 # Start command
 #    command: ["bin/acas.sh", "run", "acas", "start"]
   rservices:
-    image: mcneilco/racas-oss:${ACAS_TAG}
+    image: mcneilco/racas-oss:ACAS-633
```
Comment (brianbolt) [inline docker-compose.yml:38]: Will revert before merging

File: modules/CmpdRegBulkLoader/src/client/CmpdRegBulkLoader.coffee [hunk 4]
```diff
@@ -1203,14 +1203,14 @@ class CmpdRegBulkLoaderAppController extends Backbone.View
 			@$('.bv_bulkValSummary').hide()
 			@$('.bv_bulkRegSummary').show()
 			@setupBulkRegCmpdsSummaryController(summary[0])
-			downloadUrl = window.conf.datafiles.downloadurl.prefix + "cmpdreg_bulkload/" + encodeURIComponent(summary[1])
+			downloadUrl = window.conf.datafiles.downloadurl.prefix + encodeURIComponent(summary[1])
```
Comment (brianbolt) [inline modules/CmpdRegBulkLoader/src/client/CmpdRegBulkLoader.coffee:1206]: Service now returns full path to the file. Made a similar change to acasclient.

File: modules/CmpdRegBulkLoader/src/server/routes/CmpdRegBulkLoaderRoutes.coffee [hunk 5]
```diff
@@ -73,6 +77,13 @@ exports.getFilesToPurge = (req, resp) ->
 		baseurl = config.all.client.service.cmpdReg.persistence.fullpath+"bulkload/files"
 		serverUtilityFunctions.getFromACASServer(baseurl, resp)
 
+exports.getBulkloadFilesInternal = () ->
+	config = require '../conf/compiled/conf.js'
+	url = config.all.client.service.cmpdReg.persistence.fullpath+"bulkload/files"
+	#Fetch the url
+	response = await fetch(url, method: 'GET')
+	return await response.json()
+
```
Comment (brianbolt) [inline modules/CmpdRegBulkLoader/src/server/routes/CmpdRegBulkLoaderRoutes.coffee:86]: Needed this internal method for migration of creg files to subfolders.

File: modules/GenericDataParser/src/client/GenericDataParser.coffee [hunk 6]
```diff
@@ -8,27 +8,3 @@ class GenericDataParserController extends BasicFileValidateAndSaveController
 		@loadImagesFile = true
 		super()
 		@$('.bv_moduleTitle').html("Simple #{window.conf.experiment.label} Loader")
-
```
Comment (brianbolt) [inline modules/GenericDataParser/src/client/GenericDataParser.coffee:11]: Will revert before committing. I branched from an earlier point when this wasn't fixed and accidently committed here.

File: modules/CmpdRegBulkLoader/src/server/routes/CmpdRegBulkLoaderRoutes.coffee [hunk 7]
```diff
@@ -171,32 +182,74 @@ exports.validateCmpds = (req, resp) ->
...
-		console.log zipFilePath
+		shortZipName = path.join(exports.BULKLOAD_SUB_FOLDER, zipFileName)
+		zipFilePath = path.join(origUploadsPath, shortZipName)
+
+		console.log "Creating zip file #{zipFilePath}"
 		fstream = zip.generateNodeStream({type:"nodebuffer", streamFiles:true}).pipe(fs.createWriteStream(zipFilePath))
+		
 		fstream.on 'finish', ->
-			console.log "finished create write stream"
+			console.log "Zip file #{zipFilePath} created"
```
Comment (brianbolt) [inline modules/CmpdRegBulkLoader/src/server/routes/CmpdRegBulkLoaderRoutes.coffee:214]: This is all the new logic for moving the files to a subfolder. I should probably move this to it's own method.

File: modules/ServerAPI/src/server/Bootstrap.coffee [hunk 8]
```diff
@@ -16,15 +18,28 @@ exports.main = (callback) ->
                 callback(error)
             else
                 callback(null)
+    if config.all.server.migrateCmpdRegBulkLoaderFilesToSubfoldersOnStart? && config.all.server.migrateCmpdRegBulkLoaderFilesToSubfoldersOnStart
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/Bootstrap.coffee:21]: This is the code for the bootstrap script to call the migration to move cmpdregfiles to their new locations.

File: modules/ServerAPI/src/server/Bootstrap.coffee [hunk 9]
```diff
@@ -16,15 +18,28 @@ exports.main = (callback) ->
...
+            console.log "Created lock file at: #{lockFilePath}"
+        try
+            await fileServices.migrateCmpdRegBulkLoaderFilesToSubfolders()
+        catch err
+            console.error "Error migrating files to subfolders: #{err}"
+        finally
+            await fs.promises.unlink(lockFilePath)
+            console.log "Deleted lock file at: #{lockFilePath}"
 
-# Example install of LD client python library from local location on start
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/Bootstrap.coffee:20]: Removed this junk

File: modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee [hunk 10]
```diff
@@ -415,56 +416,37 @@ exports.getAllAuthorObjectsInternal = (callback) ->
 		callback statusCode, json
 
 exports.relocateEntityFile = (fileValue, entityCodePrefix, entityCode, callback) ->
+	console.log "relocateEntityFile"
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee:419]: Remove this unhelpful logging

File: modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee [hunk 11]
```diff
@@ -415,56 +416,37 @@ exports.getAllAuthorObjectsInternal = (callback) ->
...
+	oldPath = uploadsPath + fileValue.fileValue
 	if fileValue.comments != undefined and fileValue.comments != null
-		newPath = absEntityFolder + fileValue.comments
+		newPath = relEntityFolder + fileValue.comments
 	else
-		newPath = absEntityFolder + fileValue.fileValue
+		newPath = relEntityFolder + fileValue.fileValue
+	
+	filesToMove = [{sourceLocation: oldPath, targetLocation: newPath}]
+	movedFileResponse = await fileServices.moveDataFilesInternal(filesToMove)
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee:434]: Basically all the same logic but instead of doing all the moving here, call the "moveDataFilesInternal" method to move the files to their correct location (local or gcs).

File: modules/ServerAPI/src/server/customFunctions.R [hunk 12]
```diff
@@ -17,6 +17,73 @@ calculateTreatmemtGroupID <- function(results, inputFormat, stateGroups, resultT
...
+#' @param sourceLocations current location of files, relative from working directory or absolute
+#' @param recordedBy logged in username
+#' @param fileName name of the file to save
+#' @param entity entity object
+#' @param deleteOldFile whether to delete the old file or not
+#' @param additionalPath additional path to append to the end of the file path
+#' @return character file code or path relative from privateUploads
+#' @export
+#'
+customSourceFileMove <- function(sourceLocations, recordedBy, fileNames = NA, entityType = NA, entity = NULL,
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/customFunctions.R:31]: Previously we didn't have a "customSourceFileMove" as part of the built in ACAS so we only really suported blueimp by default.  This one is essentially to match the original customSourceFileMove signature (with a few tweaks) that R code was expecting. It also makes the function handle more than one file which was not the case before. This dramatically improves image file uploads.  This function essentially just call the ACAS move file service api and sets the path to the appropriate entity folder or subfolder of the entity folder if additionalPath is passed in (image files need to go into a subfolder of the experiment folder)

**Review by brianbolt — COMMENTED** [review]
There is a lot here and still a WIP

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 13]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+exports.moveDataFiles = (req, resp) ->
+	deleteSourceFileOnSuccess = if req.query?.deleteSourceFileOnSuccess? then req.query.deleteSourceFileOnSuccess == "true" else true
+	files = await exports.moveDataFilesInternal(req.body, deleteSourceFileOnSuccess)
+	resp.json files
+
+exports.moveDataFilesInternal = (files, deleteSourceFileOnSuccess) ->
+	return await exports.fileHandler.moveFiles(files, deleteSourceFileOnSuccess)
+
+
+exports.migrateCmpdRegBulkLoaderFilesToSubfolders = () ->
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:487]: I'm finding this method really hard to follow.

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 14]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+	console.log "About to migrate cmpdReg bulk loader files to subfolders"
+	cmpdRegBulkLoaderRoutes = require('./CmpdRegBulkLoaderRoutes.js')
+	bulkLoadFiles = await cmpdRegBulkLoaderRoutes.getBulkloadFilesInternal()
+	console.log "There are #{bulkLoadFiles.length} bulk load files in the DB"
+	bulkLoadSubFolderFiles = await exports.localFileHandler.listFiles(cmpdRegBulkLoaderRoutes.BULKLOAD_SUB_FOLDER, false)
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:492]: Ok so here we are putting together a big list of all the files already stored in the right location?
Starting with the local files?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 15]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+	console.log "There are #{bulkLoadSubFolderFiles.length} bulk load files in the local bulkload subfolder"
+	if exports.fileHandler instanceof ExternalFileHandler
+		externalBulkLoadSubFolderFiles = bulkLoadSubFolderFiles.concat(await exports.fileHandler.listFiles(cmpdRegBulkLoaderRoutes.BULKLOAD_SUB_FOLDER, false))
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:495]: Then we are looking in the external file location and checking for more properly housed files?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 16]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+			fileID = path.basename(path.dirname(file.path))
+			if !storedFilesByID[fileID]
+				storedFilesByID[fileID] = []
+			storedFilesByID[fileID].push(file)
+
+	# For each of the bulkLoadFiles in the DB check if the file is in the storedFilesByID dict
+	filesToMove = []
+	for bulkLoadFile in bulkLoadFiles
+		registeredFilesFolder = path.join(cmpdRegBulkLoaderRoutes.REGISTERED_FOLDER, bulkLoadFile.id.toString())
+		registeredStoredFiles = await exports.localFileHandler.listFiles(registeredFilesFolder)
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:509]: Now back to the local file handler for registered files? Are these the ones in the proper final location?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 17]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+		fileNameNoExtension = path.basename(bulkLoadFile.fileName, path.extname(bulkLoadFile.fileName))
+		
+		# Find all bulkLoadSubFolderFiles files that start with the fileNameNoExtension
+		matchingBulkLoadSubFolderFiles = bulkLoadSubFolderFiles.filter (file) ->
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:513]: Would this work if some bulk load files are substrings of each other?
`myfile.sdf` vs `myfile.sdf_2023-02-04_errors.sdf`

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 18]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+			path.basename(file.path).startsWith(fileNameNoExtension)
+
+		if matchingBulkLoadSubFolderFiles.length == 0 && registeredStoredFiles.length == 0
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:516]: With these variable names it's very hard for me to tell what this case means. No local files and no registered files?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 19]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+			# We move the stored DB file name
+			fileToMove = {
+				sourceLocation: file.path,
+				targetLocation: path.join(registeredFilesFolder, path.basename(file.path))
+			}
+			filesToMove.push(fileToMove)
+	
+	if filesToMove.length > 0
+		console.log "Found #{filesToMove.length} files to move"
+		await exports.localFileHandler.moveFiles(filesToMove, true)
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:535]: Does that `true` mean delete the original file after the move?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 20]
```diff
@@ -200,15 +459,134 @@ setupRoutes = (app, loginRoutes, requireLogin) ->
...
+	else 
+		console.log "No files to move"
+		
+exports.migrateFromLocalToExternalFileHandler = (lockFile) ->
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:539]: This function is way easier to follow. Makes sense to me!

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 21]
```diff
@@ -3,13 +3,229 @@ fs = require 'fs'
...
+		privateKey = config.all.server.datafiles.gcs.privateKey
+		storage = new Storage(
+			projectId: projectID,
+			credentials: {
+				client_email: clientEmail,
+				private_key: privateKey
+			}
+		)
+		uniformBucketLevelAccess = true
+		defaultAcl = [{entity: 'allUsers',role: storage.acl.READER_ROLE}]
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:168]: What does this default ACL do?

File: modules/ServerAPI/src/server/routes/FileServices.coffee [hunk 22]
```diff
@@ -3,13 +3,229 @@ fs = require 'fs'
...
+		metaData[0].contentType = metaData[0].contentType || 'application/octet-stream'
+		return {fileStream: fileStream, metaData: metaData[0]}
+
+	listFiles: (folderName, recursive) ->
+		# If recursive is true, list all files in the folder and subfolders
+
+		if folderName == undefined
+			folderName = ''
+		if recursive == undefined
+			recursive = false
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/FileServices.coffee:212]: It looks like recursive isn't used in the GCS handler. Is that intentional? Is the GCS bucket `getFiles` already recursive since it's just based on a prefix?

**Review by bffrost — COMMENTED** [review]
These are a lot of interwoven changes. I've done my best to review. Overall the different file handlers make sense to me. I had a lot of confusion around the bulk load file migration so I'm not quite ready to approve, but I'll give it another pass when you've responded to my comments @brianbolt

File: modules/ServerAPI/src/server/customFunctions.R [hunk 23]
```diff
@@ -17,6 +17,73 @@ calculateTreatmemtGroupID <- function(results, inputFormat, stateGroups, resultT
...
+  for (i in 1:length(fileNames)) {
+    if(is.na(fileNames[i])) {
+      fileNames[i] <- basename(sourceLocations[i])
+    }
+  }
+
+  # If the entity type is specified then use the entity_paths table to get the path
+  folder <- ""
+  if(!is.na(entityType)) {
+    folder <-entityFileStorePaths[entityType]
```
[the remark this answers is not in the document]
Comment (bffrost) [reply modules/ServerAPI/src/server/customFunctions.R:43]: I see you explained this on your other PR.
